openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -1,144 +1,203 @@
                                         '^.*edxapp/src.*$', '^.*call_stack_manager.*$']]
-# Variable which decides whether to track calls in the function or not. Do it by default.
+
+# Flag which decides whether to track calls in the function or not. Default True.
 TRACK_FLAG = True
 
-# List keeping track of Model classes not be tracked for special cases
-# usually cases where we know that the function is calling Model classes.
+# List keeping track of entities not to be tracked
 HALT_TRACKING = []
 
-# Module Level variables
-# dictionary which stores call stacks.
-# { "ModelClasses" : [ListOfFrames]}
+#  Dictionary which stores call logs
+# {'EntityName' : [ListOfFrames]}
 # Frames - ('FilePath','LineNumber','Context')
-# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
-#                                                    [(file,line number,context),(---,---,---)]]}
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file, line number, function name, context),(---,---,---)],
+#                                                    [(file, line number, function name, context),(---,---,---)]]}
 STACK_BOOK = collections.defaultdict(list)
 
 
-def capture_call_stack(current_model):
-    """ logs customised call stacks in global dictionary `STACK_BOOK`, and logs it.
+def capture_call_stack(entity_name):
+    """ Logs customised call stacks in global dictionary STACK_BOOK and logs it.
+
+    Arguments:
+        entity_name - entity
 
-    Args:
-        current_model - Name of the model class
     """
-    # holds temporary callstack
-    # frame[0][6:-1] -> File name along with path
-    # frame[1][6:] -> Line Number
-    # frame[2][3:] -> Context
-    temp_call_stack = [(frame[0][6:-1],
-                        frame[1][6:],
-                        frame[2][3:])
-                       for frame in [stack.replace("\n", "").strip().split(',') for stack in traceback.format_stack()]
+    # Holds temporary callstack
+    # List with each element 4-tuple(filename, line number, function name, text)
+    # and filtered with respect to regular expressions
+    temp_call_stack = [(frame[0],
+                        frame[1],
+                        frame[2],
+                        frame[3])
+                       for frame in [frames for frames in traceback.extract_stack()]
                        if not any(reg.match(frame[0]) for reg in REGULAR_EXPS)]
 
-    # avoid duplication.
-    if temp_call_stack not in STACK_BOOK[current_model] and TRACK_FLAG \
-            and not issubclass(current_model, tuple(HALT_TRACKING)):
-        STACK_BOOK[current_model].append(temp_call_stack)
-        log.info("logging new call stack for %s:\n %s", current_model, temp_call_stack)
+    def _print(frame):
+        return str('\n File ' + str(frame[0]) + ', line number ' + str(frame[1]) + ', in ' +
+                   str(frame[2]) + '\n\t' + str(frame[3]))
```

Comment:
`traceback.format_list` prints in following manner
```
['  File "spam.py", line 3, in <module>\n    spam.eggs()\n',
 '  File "eggs.py", line 42, in eggs\n    return "bacon"\n']
```
and my output is in the following way - 
```
 File /edx/app/edxapp/edx-platform/lms/djangoapps/instructor/views/api.py, line number 240, in wrapped
	return func(*args, **kwargs)
 File /edx/app/edxapp/edx-platform/lms/djangoapps/instructor/views/api.py, line number 176, in wrapped
	return func(*args, **kwargs)
 File /edx/app/edxapp/edx-platform/lms/djangoapps/instructor/views/api.py, line number 127, in wrapped
	return func(request, *args, **kwargs)
```
Python exception prints in the following way
```
Traceback (most recent call last):
  File "e.py", line 7, in <module>
    raise TypeError("Again !?!")
TypeError: Again !?!
```
which is more close to standard python exception formatting.